tree-sitter-stack-graphs/src/cli/lsp.rs

```diff
@@ -228,12 +266,52 @@ impl Backend {
             })),
         }
     }
+
+    async fn definitions(&self, reference: SourcePosition) -> Vec<SourceSpan> {
+        let mut db = match SQLiteReader::open(&self.db_path) {
+            Ok(db) => db,
+            Err(err) => {
+                self.logger
+                    .error(format!(
+                        "failed to open database {}: {}",
+                        self.db_path.display(),
+                        err
+                    ))
```

Comment:
Because all of your logging methods take in an `impl Display`, you should be able to use [`format_args!`](https://doc.rust-lang.org/std/macro.format_args.html) instead of `format!` here (and everywhere else).  Using `format!` will render each log message into a `String`, heap-allocating the necessary storage.  Using `format_args!` will instead send a _recipe_ for rendering the message, which shouldn't require any heap allocation.  That's probably not the root cause of why this is too slow, but if you're doing any kind of logging inside of hot loops it might be adding up.

(My only worry is that because your logging methods are `async` you might have to play with some lifetimes to get it to work.  Or maybe add some `Send + Sync` bounds?  Not entirely sure.)

Reply:
That's a good one, I didn't know about `format_args`, that's exactly what I'd want here!

Reply:
I tried using `format_args`, but unfortunately that doesn't play nice with either with async or thread code. The object created by `format_args` doesn't implement the necessary `Sync` or `Send` traits :/